gkr_iop/src/lib.rs

```diff
@@ -62,39 +65,43 @@ where
         &self,
         circuit: &GKRCircuit<E>,
         phase1_witness_group: &RowMajorMatrix<E::BaseField>,
+        fixed: &[Vec<E::BaseField>],
         challenges: &[E],
     ) -> (GKRCircuitWitness<'a, E>, GKRCircuitOutput<E>) {
         // layer order from output to input
+        let size = phase1_witness_group.num_instances();
         let mut layer_wits = Vec::<LayerWitness<E>>::with_capacity(circuit.layers.len() + 1);
         let phase1_witness_group = phase1_witness_group
             .to_mles()
             .into_iter()
             .map(Arc::new)
             .collect_vec();
 
-        layer_wits.push(LayerWitness::new(phase1_witness_group.clone()));
         let mut witness_mle_flatten = vec![None; circuit.n_evaluations];
 
-        // initialize a vector to store the final outputs of the GKR circuit.
-        // these outputs correspond to the evaluations at the last layer of the circuit.
-        // we preallocate the vector with exact capacity for efficiency, avoiding reallocations.
-        // the number of expected outputs is given by `circuit.n_evaluations`.
-        let mut gkr_out_well_order = Vec::with_capacity(circuit.n_evaluations);
-
         // set input to witness_mle_flatten via first layer in_eval_expr
-        let mut phase1_wit_map = HashMap::new();
-        circuit
-            .openings
-            .iter()
-            .for_each(|(phase1_wit_id, eval)| match eval {
-                EvalExpression::Zero => {}
-                EvalExpression::Single(eval_id) | EvalExpression::Linear(eval_id, _, _) => {
-                    if !phase1_wit_map.contains_key(eval_id) {
-                        phase1_wit_map.insert(*eval_id, (*phase1_wit_id, eval));
+        if let Some(first_layer) = circuit.layers.last() {
+            first_layer
+                .in_eval_expr
+                .iter()
+                .enumerate()
+                .for_each(|(index, witin)| {
+                    if index < phase1_witness_group.len() {
+                        witness_mle_flatten[*witin] = Some(phase1_witness_group[index].clone());
+                    } else {
+                        witness_mle_flatten[*witin] = Some(
+                            fixed[index - phase1_witness_group.len()]
+                                .iter()
+                                .cycle()
+                                .cloned()
+                                .take(size)
+                                .collect_vec()
+                                .into_mle()
+                                .into(),
+                        )
```

Comment:
I just change here to support shorter fixed polys. May need further modification after integration.